File: Applications/Examples/go/market_price_authentication.go
//|-----------------------------------------------------------------------------
//|            This source code is provided under the Apache 2.0 license      --
//|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
//|                See the project's LICENSE.md for details.                  --
//|           Copyright Thomson Reuters 2017. All rights reserved.            --
//|-----------------------------------------------------------------------------


// Simple example of outputting Market Price JSON data using Websockets with authentication

package main

import (
	"crypto/tls"
	"io/ioutil"
	"flag"
	"fmt"
	"log"
	"os"
	"time"
	"os/signal"
	"net/http"
	"net/url"
	"net"
	"encoding/json"
	"github.com/gorilla/websocket"
)

func main() {

	// Get command line parameters
	hostname := flag.String("hostname", "127.0.0.1", "hostname")
	port := flag.String("port", "15000", "websocket port")
	user := flag.String("user", "root", "user")
	appId := flag.String("app_id", "555", "application id")
	password := flag.String("password", "", "password")
	authHostname := flag.String("auth_hostname", "127.0.0.1", "authentication hostname")
	authPort := flag.String("auth_port", "8443", "authentication port")

	positionDefault := ""
	host, _ := os.Hostname()
	addrs, _ := net.LookupIP(host)
	for _, addr := range addrs {
		if ipv4 := addr.To4(); ipv4 != nil {
			positionDefault = fmt.Sprintf("%s",ipv4)
		}
	}

	position := flag.String("position", positionDefault, "position")

	flag.Parse()
	log.SetFlags(0)

	addr := fmt.Sprintf("%s:%s", *hostname, *port)

	interrupt := make(chan os.Signal, 1)
	signal.Notify(interrupt, os.Interrupt)

	log.Println("Sending authentication request...")
	// Send login info for authentication token
	transCfg := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	client := &http.Client{Transport: transCfg}

	authUrl := fmt.Sprintf("https://%s:%s/getToken", *authHostname, *authPort)

	authResp, err := client.PostForm(authUrl, url.Values{"username": {*user}, "password": {*password}})

	if err != nil {
		log.Println("Token request failed: ", err)
		return
	}

 	authData, _ := ioutil.ReadAll(authResp.Body)

	authResp.Body.Close()

	var authJson interface{}
	json.Unmarshal(authData, &authJson)

	authMap := authJson.(map[string]interface{})

	log.Println("RECEIVED:")
	bytesJson, _ := json.MarshalIndent(authMap, "", "  ")
	log.Println(string(bytesJson))

	if authMap["success"] == true {

		var token string
		for _, cookie := range authResp.Cookies() {
		  if cookie.Name == "AuthToken" {
			token = cookie.Value
		  }
		}

		log.Printf("Authentication Succeeded. Received AuthToken: %s\n", token)

		// Start websocket handshake
		u := url.URL{Scheme: "ws", Host: addr, Path: "/WebSocket"}
		h := http.Header{"Sec-WebSocket-Protocol": {"tr_json2"}, "Cookie": {fmt.Sprintf("AuthToken=%s;AuthPosition=%s;applicationId=%s;", token, *position, *appId)}}
		log.Printf("Connecting to WebSocket %s ...\n", u.String())

		c, _, err := websocket.DefaultDialer.Dial(u.String(), h)
		if err != nil {
			log.Fatal("WebSocket Connection Failed: ", err)
			return
		} else {
			log.Println("WebSocket successfully connected!")
		}
		defer c.Close()

		done := make(chan struct{})

		go func() {
			defer c.Close()
			defer close(done)

			// Read loop
			for {
				_, message, err := c.ReadMessage()
				if err != nil {
					log.Println("read:", err)
					return
				}

				var jsonArray []map[string]interface{}
				log.Println("RECEIVED: ")
				printJsonBytes(message)
				json.Unmarshal(message, &jsonArray)

				for _,singleMsg := range jsonArray {
					processMessage(c, singleMsg)
				}
			}
		}()

		go func() {
			for {
				time.Sleep(1*time.Second)
			}
		}()

		for {
			select {
			case <-interrupt:
				c.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
				c.Close()
				log.Println("WebSocket Closed")
				return
			}
		}
	} else {
		log.Println("Authentication failed")
	}
}

// Parse JSON message at a high level
func processMessage(c *websocket.Conn, message map[string]interface{} ) {
	switch message["Type"] {
		case "Refresh":
			if(message["Domain"] == "Login"){
				sendMarketPriceRequest(c)
			}
		case "Ping":
			sendMessage(c, []byte(`{"Type":"Pong"}`))
		default:
	}
}

// Create and send simple Market Price request
func sendMarketPriceRequest(c *websocket.Conn) {
	sendMessage(c, []byte(`{"ID":2,"Key":{"Name":"TRI.N"}}`))
}

// Helper to send bytes over WebSocket connection
func sendMessage(c *websocket.Conn, message []byte) {
	log.Println("SENT:")
	printJsonBytes(message)
	err := c.WriteMessage(websocket.TextMessage, message)
	if err != nil {
		log.Println("Send Failed:", err)
	}
}

// Output bytes as formatted JSON
func printJsonBytes(bytes []byte) {
	var dat interface{}
	json.Unmarshal(bytes, &dat)
	bytesJson, _ := json.MarshalIndent(dat, "", "  ")
	log.Println(string(bytesJson))
}

File: Applications/Examples/R/market_price_authentication.R
#|-----------------------------------------------------------------------------
#|            This source code is provided under the Apache 2.0 license      --
#|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
#|                See the project's LICENSE.md for details.                  --
#|           Copyright (C) Refinitiv 2019. All rights reserved.              --
#|-----------------------------------------------------------------------------


# Simple example of outputting Market Price JSON data using Websockets with authentication

require("websockets") # https://github.com/brettjbush/R-Websockets
require("jsonlite")
require("curl")
require("GetoptLong")

config = ""
if(.Platform$OS.type == "unix") {
  config = system("ifconfig", intern=TRUE)
  ips = config[grep("inet addr", config)]
  if(length(ips) == 0) {
    ips = config[grep("inet ", config)]
    positions = strsplit(gsub(".*inet ([[:digit:]])", "\\1", ips), " ")
    first_positions = lapply(positions, '[[', 1)
    position = first_positions[first_positions != "127.0.0.1"][[1]]
  }
  else {
    position = strsplit(gsub(".*inet addr:([[:digit:]])", "\\1", ips), " ")[[1]][[1]]
  }
} else {
  config = system("ipconfig", intern=TRUE)
  ips = config[grep("IPv4", config)]
  position = strsplit(gsub(".*? ([[:digit:]])", "\\1", ips), " ")[[1]][[1]]
}

# Global Default Variables
hostname = "127.0.0.1"
port = "15000"
user = "root"
app_id = "555"
auth_token = ""
auth_hostname = "127.0.0.1"
auth_port = "8443"
password = ""

# Get command line parameters
GetoptLong(
  "hostname=s","",
  "port=s","",
  "user=s","",
  "app_id=s","",
  "position=s","",
  "password=s","",
  "auth_hostname=s","",
  "auth_port=s",""
)

# Send login info for authentication token
cat("Sending authentication request...\n")
content = paste("username=", user, "&password=", password, sep="")
h <- new_handle(copypostfields = content)
handle_setheaders(h,
  "Content-Type" = "application/x-www-form-urlencoded"
)
handle_setopt(h, ssl_verifypeer = FALSE, ssl_verifyhost = FALSE)
auth_url = paste("https://", auth_hostname, ":", auth_port, "/getToken", sep="")
req <- curl_fetch_memory(auth_url, handle = h)
res_headers = parse_headers(req$headers)
auth_json_string = rawToChar(req$content)
auth_json = fromJSON(auth_json_string)

cat("RECEIVED:\n")
cat(toJSON(auth_json, pretty=TRUE, auto_unbox=TRUE))
cat("\n")

if (auth_json$success) {
  auth_token = gsub(".*AuthToken=([[:alnum:]]*);.*", "\\1", grep('AuthToken', res_headers, value=TRUE))
  cat("Authentication Succeeded. Received AuthToken: ")
  cat(auth_token)
  cat("\n")

  # Start websocket handshake
  ws_address = paste("ws://", hostname, ":", port, "/WebSocket", sep="")
  cat(paste("Connecting to WebSocket", ws_address, "...\n"))
  header_entry = list(name="Cookie", value=paste("AuthToken=", auth_token, ";AuthPosition=", position, ";applicationId=", app_id,";", sep=""))
  headers_list = list(header_entry)
  con = websocket(ws_address, port=as.integer(port), subprotocol="tr_json2", headers=headers_list, version=13)

  # Create and send simple Market Price request
  send_market_price_request = function(con) {
    mp_req_json_string = "{\"ID\":2,\"Key\":{\"Name\":\"TRI.N\"}}"
    mp_req_json = fromJSON(mp_req_json_string)
    websocket_write(mp_req_json_string, con)
    cat("SENT:\n")
    cat(toJSON(mp_req_json, pretty=TRUE, auto_unbox=TRUE))
    cat("\n")
  }

  # Parse at high level and output JSON of message
  process_message = function(con, message_json) {
    message_type = message_json$Type

    if(message_type == "Refresh") {
  	message_domain = message_json$Domain
  	if(!is.null(message_domain)) {
  	  if(message_domain == "Login") {
  	    send_market_price_request(con)
  	  }
    }
    } else if (message_type == "Ping") {
      pong_json_string = "{\"Type\":\"Pong\"}"
      pong_json = fromJSON(pong_json_string)
      websocket_write(pong_json_string, con)
      cat("SENT:\n")
      cat(toJSON(pong_json, pretty=TRUE, auto_unbox=TRUE))
      cat("\n")
      }
  }

  # Called when handshake is complete and websocket is open, send login
  e = function(WS) {
    cat("WebSocket successfully connected!\n")
  }
  set_callback("established", e, con)

  # Called when message received, parse message into JSON for processing
  r = function(DATA,WS,...) {
    message = rawToChar(DATA)
    jsonArray = fromJSON(message, simplifyDataFrame=FALSE)

    cat("RECEIVED:\n")
    cat(toJSON(jsonArray, pretty=TRUE, auto_unbox=TRUE))
    cat("\n")

    for (singleMsg in jsonArray)
       process_message(con, singleMsg)
  }
  set_callback("receive", r, con)

  # Called when websocket is closed
  c = function(WS) {
    cat("WebSocket Closed\n")
  }
  set_callback("closed", c, con)

  while (TRUE) {
    service(con)
  }
} else {
  cat("Authentication failed")
}
